### PR TITLE
Encrypt stored keys

### PR DESCRIPTION
**Pull request by rvolosatovs**
#### Summary

Closes https://github.com/TheThingsNetwork/lorawan-stack/issues/847

#### Changes

- Encrypted keys stored via registries using device KEK
- Fix AS and JS registries not testing the device actually being set

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, database and configuration, according to the stability commitments in `README.md`.
- [x] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [x] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

**Comment by rvolosatovs** [comment]
Blocked by https://github.com/TheThingsNetwork/lorawan-stack/pull/1550

File: pkg/joinserver/joinserver.go [hunk 1]
```diff
@@ -54,6 +54,7 @@ type Config struct {
 	Devices         DeviceRegistry      `name:"-"`
 	Keys            KeyRegistry         `name:"-"`
 	JoinEUIPrefixes []types.EUI64Prefix `name:"join-eui-prefix" description:"JoinEUI prefixes handled by this JS"`
+	DeviceKEKLabel  string              `name:"device-kek-label" description:"Label of KEK used to encrypt device keys"`
```
Comment (johanstokking) [inline pkg/joinserver/joinserver.go]: Idem

File: pkg/networkserver/config.go [hunk 2]
```diff
@@ -35,6 +35,7 @@ type Config struct {
 	DownlinkPriorities  DownlinkPriorityConfig `name:"downlink-priorities" description:"Downlink message priorities"`
 	DefaultMACSettings  MACSettingConfig       `name:"default-mac-settings" description:"Default MAC settings to fallback to if not specified by device, band or frequency plan"`
 	Interop             config.InteropClient   `name:"interop" description:"Interop client configuration"`
+	DeviceKEKLabel      string                 `name:"device-kek-label" description:"Label of KEK used to encrypt device keys"`
```
Comment (johanstokking) [inline pkg/networkserver/config.go]: Idem

File: pkg/joinserver/grpc_deviceregistry.go [hunk 3]
```diff
@@ -193,26 +187,75 @@ func (srv jsEndDeviceRegistryServer) Set(ctx context.Context, req *ttnpb.SetEndD
...
 		}
-		// TODO: Encrypt keys using component-local KEK(https://github.com/TheThingsNetwork/lorawan-stack/issues/847).
+	}
+
+	if req.EndDevice.JoinEUI == nil || req.EndDevice.JoinEUI.IsZero() {
+		return nil, errNoJoinEUI
+	}
+	if req.EndDevice.DevEUI == nil || req.EndDevice.DevEUI.IsZero() {
+		return nil, errNoDevEUI
+	}
```
Comment (johanstokking) [inline pkg/joinserver/grpc_deviceregistry.go]: Why change order? These are cheap checks, while `RequireApplication()` does RPC calls

File: pkg/joinserver/grpc_deviceregistry.go [hunk 4]
```diff
@@ -193,26 +187,75 @@ func (srv jsEndDeviceRegistryServer) Set(ctx context.Context, req *ttnpb.SetEndD
...
+
+	sets := append(req.FieldMask.Paths[:0:0], req.FieldMask.Paths...)
+	if ttnpb.HasAnyField(req.FieldMask.Paths, "root_keys.root_key_id") && req.EndDevice.RootKeys.GetRootKeyID() == "" {
+		return nil, errInvalidFieldValue.WithAttributes("field", "root_keys.root_key_id")
```
Comment (johanstokking) [inline pkg/joinserver/grpc_deviceregistry.go]: This will break device claiming

File: pkg/joinserver/grpc_deviceregistry.go [hunk 5]
```diff
@@ -193,26 +187,75 @@ func (srv jsEndDeviceRegistryServer) Set(ctx context.Context, req *ttnpb.SetEndD
...
+	}
+	if ttnpb.HasAnyField(req.FieldMask.Paths, "root_keys.app_key.key") {
+		if req.EndDevice.RootKeys.GetAppKey().GetKey().IsZero() {
+			return nil, errInvalidFieldValue.WithAttributes("field", "root_keys.app_key.key")
```
Comment (johanstokking) [inline pkg/joinserver/grpc_deviceregistry.go]: Idem

File: pkg/networkserver/grpc_deviceregistry.go [hunk 6]
```diff
@@ -234,43 +234,9 @@ func (ns *NetworkServer) Get(ctx context.Context, req *ttnpb.GetEndDeviceRequest
...
-	}
-	if ttnpb.HasAnyField(req.FieldMask.Paths, "lorawan_phy_version") {
-		if err := req.EndDevice.LoRaWANPHYVersion.Validate(); err != nil {
-			return nil, errInvalidFieldValue.WithAttributes("field", "lorawan_phy_version").WithCause(err)
-		}
-	}
-
-	if err := rights.RequireApplication(ctx, req.EndDevice.ApplicationIdentifiers, ttnpb.RIGHT_APPLICATION_DEVICES_WRITE); err != nil {
+func (ns *NetworkServer) Set(ctx context.Context, req *ttnpb.SetEndDeviceRequest) (dev *ttnpb.EndDevice, err error) {
+	if err = rights.RequireApplication(ctx, req.EndDevice.ApplicationIdentifiers, ttnpb.RIGHT_APPLICATION_DEVICES_WRITE); err != nil {
```
Comment (johanstokking) [inline pkg/networkserver/grpc_deviceregistry.go]: Why expensive check first?

File: pkg/joinserver/grpc_deviceregistry.go [hunk 7]
```diff
@@ -193,26 +194,67 @@ func (srv jsEndDeviceRegistryServer) Set(ctx context.Context, req *ttnpb.SetEndD
...
-	dev, err := srv.JS.devices.SetByID(ctx, req.EndDevice.ApplicationIdentifiers, req.EndDevice.DeviceID, req.FieldMask.Paths, func(dev *ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error) {
-		sets := req.FieldMask.Paths
-		if dev == nil {
-			evt = evtCreateEndDevice(ctx, req.EndDevice.EndDeviceIdentifiers, nil)
-		} else {
+	dev, err = srv.JS.devices.SetByID(ctx, req.EndDevice.ApplicationIdentifiers, req.EndDevice.DeviceID, req.FieldMask.Paths, func(dev *ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error) {
+		if dev != nil {
 			evt = evtUpdateEndDevice(ctx, req.EndDevice.EndDeviceIdentifiers, req.FieldMask.Paths)
-			if err := ttnpb.ProhibitFields(req.FieldMask.Paths, "ids.dev_addr"); err != nil {
+			if err := ttnpb.ProhibitFields(sets,
```
Comment (htdvisser) [inline pkg/joinserver/grpc_deviceregistry.go]: This is unnecessary

File: pkg/networkserver/grpc_deviceregistry.go [hunk 8]
```diff
@@ -234,43 +234,43 @@ func (ns *NetworkServer) Get(ctx context.Context, req *ttnpb.GetEndDeviceRequest
...
+			return nil, errInvalidFieldValue.WithAttributes("field", "lorawan_phy_version").WithCause(err)
+		}
+	}
+	if ttnpb.HasAnyField(req.FieldMask.Paths, "lorawan_version") {
+		if err := req.EndDevice.LoRaWANVersion.Validate(); err != nil {
+			return nil, errInvalidFieldValue.WithAttributes("field", "lorawan_version").WithCause(err)
+		}
+	}
+
 	if ttnpb.HasAnyField(req.FieldMask.Paths, "session.dev_addr") && (req.EndDevice.Session == nil || req.EndDevice.Session.DevAddr.IsZero()) {
```
Comment (htdvisser) [inline pkg/networkserver/grpc_deviceregistry.go]: `HasAnyField` also checks for prefixes, so you could eliminate multiple `req.EndDevice.Session == nil` checks below with a single

```go
if ttnpb.HasAnyField(req.FieldMask.Paths, "session") && req.EndDevice.Session == nil {
  ...
}
```

File: pkg/networkserver/grpc_deviceregistry.go [hunk 9]
```diff
@@ -234,43 +234,43 @@ func (ns *NetworkServer) Get(ctx context.Context, req *ttnpb.GetEndDeviceRequest
...
+
+	if ttnpb.HasAnyField(req.FieldMask.Paths, "multicast") && ttnpb.HasAnyField(req.FieldMask.Paths, "supports_join") && req.EndDevice.Multicast && req.EndDevice.SupportsJoin {
+		return nil, errInvalidFieldValue.WithAttributes("field", "supports_join")
 	}
 
-	if err := rights.RequireApplication(ctx, req.EndDevice.ApplicationIdentifiers, ttnpb.RIGHT_APPLICATION_DEVICES_WRITE); err != nil {
+	if err = rights.RequireApplication(ctx, req.EndDevice.ApplicationIdentifiers, ttnpb.RIGHT_APPLICATION_DEVICES_WRITE); err != nil {
 		return nil, err
 	}
 	if ttnpb.HasAnyField(req.FieldMask.Paths,
```
Comment (htdvisser) [inline pkg/networkserver/grpc_deviceregistry.go]: Also here you can simplify by leveraging the fact that `HasAnyField` checks for prefixes.

File: pkg/networkserver/grpc_deviceregistry.go (hunk 8 above)
Comment (rvolosatovs) [reply pkg/networkserver/grpc_deviceregistry.go]: That is not true
```
diff --git a/pkg/ttnpb/fieldmask_utils_test.go b/pkg/ttnpb/fieldmask_utils_test.go
index e51836044..fe6f3f0b8 100644
--- a/pkg/ttnpb/fieldmask_utils_test.go
+++ b/pkg/ttnpb/fieldmask_utils_test.go
@@ -101,6 +101,13 @@ func TestHasAnyField(t *testing.T) {
 	a.So(HasAnyField(requested, "x", "b.c.d"), should.BeTrue)
 	a.So(HasAnyField(requested, "d"), should.BeTrue)
 	a.So(HasAnyField(requested, "d.e", "b"), should.BeTrue)
+	a.So(HasAnyField(requested, "b"), should.BeTrue)
+	a.So(HasAnyField([]string{
+		"session.keys",
+		"session.dev_addr",
+	},
+		"session",
+	), should.BeTrue)
 }
 
 func TestFlattenPaths(t *testing.T) {
```

File: pkg/networkserver/grpc_deviceregistry.go (hunk 9 above)
Comment (rvolosatovs) [reply pkg/networkserver/grpc_deviceregistry.go]: Unfortunately, not https://github.com/TheThingsNetwork/lorawan-stack/pull/1591#discussion_r348566878

**Pull request merged**